data_generate.py

```diff
@@ -0,0 +1,157 @@
+import numpy as np
+from scipy.sparse import vstack, csr_matrix
+import gym, recogym
+import pandas as pd
+import math
+from copy import deepcopy
+from tqdm import tqdm, trange
+from recogym import env_1_args, Configuration
+from recogym.agents import RandomAgent, random_args
+import sys
+
+def train_data(data, num_products, va_ratio=0.2, te_ratio=0.2):
+        data = pd.DataFrame().from_dict(data)
+
+        features = []
+        actions = []
+        pss = []
+        deltas = []
+        set_flags = []
+
+        for user_id in tqdm(data['u'].unique()):
+            tmp_feature = []
+            tmp_action = []
+            tmp_ps = []
+            tmp_delta = []
+            tmp_set_flag = []
+
+            views = np.zeros((0, num_products))
+            history = np.zeros((0, 1))
+            for _, user_datum in data[data['u'] == user_id].iterrows():
+                assert (not math.isnan(user_datum['t']))
+                if user_datum['z'] == 'organic':
+                    assert (math.isnan(user_datum['a']))
+                    assert (math.isnan(user_datum['c']))
+                    assert (not math.isnan(user_datum['v']))
+
+                    view = int(user_datum['v'])
+
+                    tmp_view = np.zeros(num_products)
+                    tmp_view[view] = 1
+
+                    # Append the latest view at the beginning of all views.
+                    views = np.append(tmp_view[np.newaxis, :], views, axis = 0)
+                    history = np.append(np.array([user_datum['t']])[np.newaxis, :], history, axis = 0)
+                else:
+                    assert (user_datum['z'] == 'bandit')
+                    assert (not math.isnan(user_datum['a']))
+                    assert (not math.isnan(user_datum['c']))
+                    assert (math.isnan(user_datum['v']))
+
+                    action = int(user_datum['a'])
+                    delta = int(user_datum['c'])
+                    ps = user_datum['ps']
+                    time = user_datum['t']
+
+                    train_views = views
+
+                    feature = np.sum(train_views, axis = 0)
+                    feature = feature/np.linalg.norm(feature)
```

Comment:
Have we decided this is the way to do this?

Reply:
One of the other options is to set the feature to be one as long as it is greater than zero.

Reply:
I just choose one of these two options randomly. But I think we can keep this now as it is not the main factor for our exps.